### PR TITLE
Ensure that  textDirection is always passed to ShapeBorder.getOuterPath

### LINKED ISSUE #21506: Button with Border trips assert in `Border.getOuterPath`
After upgrading to Flutter beta v0.7.3, I am now seeing an assert when Buttons that have a Border attached are tapped. This issue did not occur on beta v0.6.0 or v0.5.1, so it appears to be a regression.

```
flutter: The textDirection argument to Border.getOuterPath must not be null.
flutter: 'package:flutter/src/painting/box_border.dart': Failed assertion: line 181 pos 12: 'textDirection !=
```

## Steps to Reproduce

Create an app with the following main.dart: https://gist.github.com/kibantony/5e5d7a6c3861bdf6c083299808b06803

Run the app (I was using the iOS simulator).
Click on the `Test` button.
Observe the debug console output

## Logs

```
Launching lib/main.dart on iPhone 8 in debug mode...
Xcode build done.
flutter: ══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
flutter: The following assertion was thrown during paint():
flutter: The textDirection argument to Border.getOuterPath must not be null.
flutter: 'package:flutter/src/painting/box_border.dart': Failed assertion: line 181 pos 12: 'textDirection !=
flutter: null'
flutter:
flutter: Either the assertion indicates an error in the framework itself, or we should provide substantially
flutter: more information in this error message to help you determine and fix the underlying cause.
flutter: In either case, please report this assertion by filing a bug on GitHub:
flutter:   https://github.com/flutter/flutter/issues/new
flutter:
flutter: When the exception was thrown, this was the stack:
flutter: #2      BoxBorder.getOuterPath (package:flutter/src/painting/box_border.dart:181:12)
flutter: #3      InkSplash.paintFeature (package:flutter/src/material/ink_splash.dart:209:39)
flutter: #4      InkFeature._paint (package:flutter/src/material/material.dart:562:5)
flutter: #5      _RenderInkFeatures.paint (package:flutter/src/material/material.dart:469:20)
flutter: #6      RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #7      PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #8      _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #9      RenderCustomPaint.paint (package:flutter/src/rendering/custom_paint.dart:561:11)
flutter: #10     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #11     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #12     _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #13     RenderPhysicalShape.paint.<anonymous closure> (package:flutter/src/rendering/proxy_box.dart:1822:86)
flutter: #14     ClipContext._clipAndPaint (package:flutter/src/painting/clip.dart:29:12)
flutter: #15     ClipContext.clipPathAndPaint (package:flutter/src/painting/clip.dart:41:5)
flutter: #16     RenderPhysicalShape.paint (package:flutter/src/rendering/proxy_box.dart:1822:17)
flutter: #17     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #18     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #19     _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #20     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #21     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #22     RenderShiftedBox.paint (package:flutter/src/rendering/shifted_box.dart:70:15)
flutter: #23     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #24     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #25     _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #26     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #27     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #28     RenderShiftedBox.paint (package:flutter/src/rendering/shifted_box.dart:70:15)
flutter: #29     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #30     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #31     _RenderCustomMultiChildLayoutBox&RenderBox&ContainerRenderObjectMixin&RenderBoxContainerDefaultsMixin.defaultPaint (package:flutter/src/rendering/box.dart:2267:15)
flutter: #32     RenderCustomMultiChildLayoutBox.paint (package:flutter/src/rendering/custom_layout.dart:360:5)
flutter: #33     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #34     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #35     _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #36     _RenderInkFeatures.paint (package:flutter/src/material/material.dart:472:11)
flutter: #37     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #38     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #39     _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #40     PaintingContext.pushLayer (package:flutter/src/rendering/object.dart:305:12)
flutter: #41     RenderPhysicalModel.paint (package:flutter/src/rendering/proxy_box.dart:1689:17)
flutter: #42     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #43     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #44     _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #45     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #46     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:131:13)
flutter: #47     _RenderProxyBox&RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:126:15)
flutter: #48     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2017:7)
flutter: #49     PaintingContext.repaintCompositedChild (package:flutter/src/rendering/object.dart:111:11)
flutter: #50     PipelineOwner.flushPaint (package:flutter/src/rendering/object.dart:784:29)
flutter: #51     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:272:19)
flutter: #52     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:654:13)
flutter: #53     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&SemanticsBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:208:5)
flutter: #54     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:990:15)
flutter: #55     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:930:9)
flutter: #56     _WidgetsFlutterBinding&BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:842:5)
flutter: #57     _invoke (dart:ui/hooks.dart:128:13)
flutter: #58     _drawFrame (dart:ui/hooks.dart:117:3)
flutter: (elided 2 frames from class _AssertionError)
flutter:
flutter: The following RenderObject was being processed when the exception was fired:
flutter:   _RenderInkFeatures#f9c18 relayoutBoundary=up7
flutter:   creator: _InkFeatures-[GlobalKey#f29ab ink renderer] ←
flutter:   NotificationListener<LayoutChangedNotification> ← CustomPaint ← _ShapeBorderPaint ← PhysicalShape
flutter:   ← _MaterialInterior ← Material ← ConstrainedBox ← _InputPadding ← Semantics ← RawMaterialButton ←
flutter:   RaisedButton ← ⋯
flutter:   parentData: <none> (can use size)
flutter:   constraints: BoxConstraints(88.0<=w<=375.0, 36.0<=h<=591.0)
flutter:   size: Size(88.0, 36.0)
flutter: This RenderObject had the following descendants (showing up to depth 5):
flutter:   RenderSemanticsGestureHandler#6af1e relayoutBoundary=up8
flutter:     RenderPointerListener#c5bf5 relayoutBoundary=up9
flutter:       RenderPadding#f95e5 relayoutBoundary=up10
flutter:         RenderPositionedBox#b8e1f relayoutBoundary=up11
flutter:           RenderParagraph#04424 relayoutBoundary=up12
flutter: ════════════════════════════════════════════════════════════════════════════════════════════════════
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
flutter: Another exception was thrown: The textDirection argument to Border.getOuterPath must not be null.
```

## flutter doctor -v

```
[✓] Flutter (Channel beta, v0.7.3, on Mac OS X 10.13.6 17G65, locale en-US)
    • Flutter version 0.7.3 at /Users/tony/afero/flutter
    • Framework revision 3b309bda07 (9 days ago), 2018-08-28 12:39:24 -0700
    • Engine revision af42b6dc95
    • Dart version 2.1.0-dev.1.0.flutter-ccb16f7282

[✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at /Users/tony/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-27, build-tools 27.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 9.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.4.1, Build version 9F2000
    • ios-deploy 1.9.2
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 27.1.1
    • Dart plugin version 173.4700
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1024-b01)

[✓] VS Code (version 1.27.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 2.18.0

[✓] Connected devices (1 available)
    • iPhone 8 • EEF6AF6E-7CBB-464D-9115-131D04A6A5B6 • ios • iOS 11.4 (simulator)

• No issues found!
```

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fix https://github.com/flutter/flutter/issues/21506

### CODE REVIEW

File: packages/flutter/test/material/ink_splash_test.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  // Regression test for https://github.com/flutter/flutter/issues/21506.
+  testWidgets('InkSplash receives textDirection', (WidgetTester tester) async {
+    await tester.pumpWidget(new MaterialApp(
+      home: new Scaffold(
+      appBar: new AppBar(title: const Text('Button Border Test')),
+      body: new Center(
+        child: new RaisedButton(
+          child: const Text('Test'),
+          onPressed: () {},
+          shape: Border.all(
```
Comment (Hixie) [inline packages/flutter/test/material/ink_splash_test.dart]: missing a `new`

**Comment by Hixie** [comment]
Do all the widgets to which you added calls to `Directionality.of` already call debugCheckHasDirectionality?

File: packages/flutter/test/material/ink_splash_test.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/test/material/ink_splash_test.dart]: It seems weird that `new Border.all` ends up requiring a textDirection. That one should be direction-agnostic.

File: packages/flutter/lib/src/material/ink_ripple.dart [hunk 2]
```diff
@@ -245,7 +249,7 @@ class InkRipple extends InteractiveInkFeature {
     if (_clipCallback != null) {
       final Rect rect = _clipCallback();
       if (_customBorder != null) {
-        canvas.clipPath(_customBorder.getOuterPath(rect));
+        canvas.clipPath(_customBorder.getOuterPath(rect, textDirection: _textDirection));
```
Comment (kibantony) [inline packages/flutter/lib/src/material/ink_ripple.dart]: Is it understood why a missing `@required` parameter wasn't caught by the various build checks & analysis? For that matter why is `textDirection` a named parameter in this case?

Comment (jonahwilliams) [reply packages/flutter/lib/src/material/ink_ripple.dart]: The textDirection isn't `@required` in ShapeBorder, and that is the static type of `_customBorder` here.

**Comment by HansMuller** [comment]
InkHighlight and OutlineButton have the same problem - getOutlineRect() calls without a textDirection parameter

File: packages/flutter/lib/src/material/ink_ripple.dart [hunk 3]
```diff
@@ -120,11 +121,13 @@ class InkRipple extends InteractiveInkFeature {
     ShapeBorder customBorder,
     double radius,
     VoidCallback onRemoved,
+    TextDirection textDirection,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/ink_ripple.dart]: This class can fail if textDirection is null - why isn't this parameter @required and asserted to be non-null?

File: packages/flutter/lib/src/material/ink_splash.dart [hunk 4]
```diff
@@ -57,6 +57,7 @@ class _InkSplashFactory extends InteractiveInkFeatureFactory {
     ShapeBorder customBorder,
     double radius,
     VoidCallback onRemoved,
+    TextDirection textDirection,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/ink_splash.dart]: This class can fail if textDirection is null - why isn't this parameter @required and asserted to be non-null?

**Comment by jonahwilliams** [comment]
@HansMuller what part of the OutlineButton is this? I passed the text direction through the outline border's paint method

File: packages/flutter/lib/src/material/ink_splash.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/material/ink_splash.dart]: Added assert to constructor and required to create/constructor

File: packages/flutter/lib/src/material/ink_ripple.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter/lib/src/material/ink_ripple.dart]: Added assert to constructor and required to create/constructor

**Comment by HansMuller** [comment]
The _OutlineBorder change you made (https://github.com/flutter/flutter/pull/21532/files#diff-9d493d843ee1c772332e1f36c4eb6387R500) is what I was getting at.

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by Hixie** [comment]
FWIW, `textDirection` isn't always `@required` because there's cases where these APIs can be used in direction-agnostic ways. For example, many shapes (e.g. circles) don't care about direction, and so you can use those shapes even if you don't have a direction.

**Pull request merged**